File: versao-java/src/main/java/api/produtos/java/controller/ProdutoController.java
package api.produtos.java.controller;

import api.produtos.java.entity.Produto;
import api.produtos.java.repository.ProdutoRepository;
import org.springframework.web.bind.annotation.*;

import java.util.List;

import static org.springframework.http.HttpStatus.CREATED;
import static org.springframework.http.HttpStatus.OK;

@RestController
@RequestMapping("/produtos")
public class ProdutoController {
    private ProdutoRepository repository;

    public ProdutoController(ProdutoRepository repository) {
        this.repository = repository;
    }

    @PostMapping
    @ResponseStatus(CREATED)
    public Produto criar(@RequestBody Produto produto) {
        produto.setId(null);
        return repository.save(produto);
    }

    @GetMapping
    @ResponseStatus(OK)
    public List<Produto> buscar() {
        return repository.findAll();
    }

    @PutMapping("/{id}")
    @ResponseStatus(OK)
    public Produto atualizar(@PathVariable Long id,
                             @RequestBody Produto produto) {
        produto.setId(id);
        return repository.save(produto);
    }

    @DeleteMapping("/{id}")
    @ResponseStatus(OK)
    public Produto deletar(@PathVariable Long id) {
        Produto produto = repository.findById(id).orElseThrow();
        repository.deleteById(id);
        return produto;
    }
}

File: versao-kotlin/src/main/kotlin/api/produtos/kotlin/controller/ProdutoController.kt
package api.produtos.kotlin.controller

import api.produtos.kotlin.entity.Produto
import api.produtos.kotlin.repository.ProdutoRepository
import org.springframework.http.HttpStatus.CREATED
import org.springframework.http.HttpStatus.OK
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/produtos")
class ProdutoController(val repository: ProdutoRepository) {
    @PostMapping
    @ResponseStatus(CREATED)
    fun criar(@RequestBody produto: Produto) =
        repository.save(produto.apply { id = null })

    @GetMapping
    @ResponseStatus(OK)
    fun buscar() = repository.findAll()

    @PutMapping("/{id}")
    @ResponseStatus(OK)
    fun atualizar(
        @PathVariable id: Long,
        @RequestBody produto: Produto
    ) =
        repository.save(produto.also { it.id = id })

    @DeleteMapping("/{id}")
    @ResponseStatus(OK)
    fun deletar(@PathVariable id: Long) =
        repository.findById(id).orElseThrow()
            .also { repository.deleteById(id) }
}